### PR TITLE
2940 hosted proxy

### PR DESCRIPTION
**Pull request by cameronmwall**
Issue: https://issues.redhat.com/browse/ACM-2940

### CODE REVIEW

**Review by JakobGray — CHANGES_REQUESTED** [review]
(comment from https://github.com/stolostron/backplane-operator/pull/390) Can you provide a list of commands to run to verify everything is deployed as it should be? For example:
```
oc get deploy cluster-manager-webhook-konnectivity-agent -n <mceNS>
oc get service cluster-manager-work-webhook -n <mceNS>
KUBECONFIG=<hosted-kubeconfig> oc get service .....
```

I tried the oc api-resources call and nothing stood out to me before or after the changes were applied so I'm not sure if anything happened.

I tried running this code and the hosted MCE couldn't get off the ground because of this error.
```
2023-03-28T18:00:40Z	INFO	error rendering chart: konnectivity-cm-webhook-agent
2023-03-28T18:00:40Z	INFO	template: konnectivity-cm-webhook-agent/templates/cm-webhook-agent.yaml:70:25: executing "konnectivity-cm-webhook-agent/templates/cm-webhook-agent.yaml" at <.Values.global.imageOverrides.apiserver_network_policy>: map has no entry for key "apiserver_network_policy"
2023-03-28T18:00:40Z	INFO	template: konnectivity-cm-webhook-agent/templates/cm-webhook-agent.yaml:70:25: executing "konnectivity-cm-webhook-agent/templates/cm-webhook-agent.yaml" at <.Values.global.imageOverrides.apiserver_network_policy>: map has no entry for key "apiserver_network_policy"	{"controller": "multiclusterengine", "controllerGroup": "multicluster.openshift.io", "controllerKind": "MultiClusterEngine", "MultiClusterEngine": {"name":"wombo"}, "namespace": "", "name": "wombo", "reconcileID": "8bc31536-bf60-44c2-9d68-16b6974ef36d"}
```
After you've made your changes please test them in a running environment before requesting another code review.

**Comment by cameronmwall** [comment]
> (comment from #390) Can you provide a list of commands to run to verify everything is deployed as it should be? For example:
> 
> ```
> oc get deploy cluster-manager-webhook-konnectivity-agent -n <mceNS>
> oc get service cluster-manager-work-webhook -n <mceNS>
> KUBECONFIG=<hosted-kubeconfig> oc get service .....
> ```
> 
> I tried the oc api-resources call and nothing stood out to me before or after the changes were applied so I'm not sure if anything happened.
> 
> I tried running this code and the hosted MCE couldn't get off the ground because of this error.
> 
> ```
> 2023-03-28T18:00:40Z	INFO	error rendering chart: konnectivity-cm-webhook-agent
> 2023-03-28T18:00:40Z	INFO	template: konnectivity-cm-webhook-agent/templates/cm-webhook-agent.yaml:70:25: executing "konnectivity-cm-webhook-agent/templates/cm-webhook-agent.yaml" at <.Values.global.imageOverrides.apiserver_network_policy>: map has no entry for key "apiserver_network_policy"
> 2023-03-28T18:00:40Z	INFO	template: konnectivity-cm-webhook-agent/templates/cm-webhook-agent.yaml:70:25: executing "konnectivity-cm-webhook-agent/templates/cm-webhook-agent.yaml" at <.Values.global.imageOverrides.apiserver_network_policy>: map has no entry for key "apiserver_network_policy"	{"controller": "multiclusterengine", "controllerGroup": "multicluster.openshift.io", "controllerKind": "MultiClusterEngine", "MultiClusterEngine": {"name":"wombo"}, "namespace": "", "name": "wombo", "reconcileID": "8bc31536-bf60-44c2-9d68-16b6974ef36d"}
> ```
> 
> After you've made your changes please test them in a running environment before requesting another code review.

```
oc get mce
oc get deployment -n <hosted cluster manager namespace>
oc get service -n <hosted target namespace>
oc get deployment-n <hosted target namespace>
oc get networkpolicy konnectivity-server-from-hub-ns -n <hosted cluster namespace>
```

We can check that the MCE has come up to available. In the hosted cluster manager namespace, we can check that the registration and work web hook deployments are up and running. In the hosted target namespace we can look for the services. Then, we can check to make sure that the cluster IPs from the services in the hosted target namespace match the konnectivity server deployment in the hosted target namespace

File: controllers/hosted.go [hunk 1]
```diff
@@ -366,26 +374,141 @@ func (r *MultiClusterEngineReconciler) ensureHostedClusterManager(ctx context.Co
...
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "error getting registration webhook service")
+		return ctrl.Result{}, err
+	}
+	err2 := r.Client.Get(ctx, workWebhookServiceNN, workWebhookService)
+	if err2 != nil && !apierrors.IsNotFound(err2) {
+		log.Error(err2, "error getting work webhook service")
+		return ctrl.Result{}, err2
+	}
+	addresses := []string{registrationWebhookService.Spec.ClusterIP, workWebhookService.Spec.ClusterIP}
```
[the remark this answers is not in the document]
Comment (cameronmwall) [reply controllers/hosted.go:445]: From my testing, there is no risk of panicking. If the services are not found, the reconciler simply moves on and uses the default values for IP

**Comment by JakobGray** [comment]
> We can check that the MCE has come up to available. In the hosted cluster manager namespace, we can check that the registration and work web hook deployments are up and running. In the hosted target namespace we can look for the services. Then, we can check to make sure that the cluster IPs from the services in the hosted target namespace match the konnectivity server deployment in the hosted target namespace

@cameronmwall When you say hosted target namespace do you mean the MCE targetNS on the hosting cluster, or a namespace on the hosted cluster?

**Review by JakobGray — COMMENTED** [review]
I think everything is now getting deployed. As a sanity test I tried the webhook and got an error. Is this a valid way to test the webhook proxying?

```
KUBECONFIG=${HOSTED_CLUSTER_KUBECONFIG}.kubeconfig oc apply -f work-example.yaml
Error from server (InternalError): error when creating "work-example.yaml": Internal error occurred: failed calling webhook "manifestworkvalidators.admission.work.open-cluster-management.io": failed to call webhook: Post "https://cluster-manager-work-webhook.wombo-cluster-manager.svc:443/validate-work-open-cluster-management-io-v1-manifestwork?timeout=10s": EOF
```

**Pull request closed without merging**